File: 2024/11/Go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func processStonesOptimized(stones []int, blinks int) int {
	counts := make(map[int]int)
	for _, stone := range stones {
		counts[stone]++
	}

	for i := 0; i < blinks; i++ {
		newCounts := make(map[int]int)

		for stone, count := range counts {
			if stone == 0 {
				newCounts[1] += count
			} else if len(strconv.Itoa(stone))%2 == 0 {
				numStr := strconv.Itoa(stone)
				mid := len(numStr) / 2
				left, _ := strconv.Atoi(numStr[:mid])
				right, _ := strconv.Atoi(numStr[mid:])
				newCounts[left] += count
				newCounts[right] += count
			} else {
				newStone := stone * 2024
				newCounts[newStone] += count
			}
		}
		counts = newCounts
	}
	total := 0
	for _, count := range counts {
		total += count
	}
	return total
}

func main() {
	data, err := ioutil.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}
	stonesStr := strings.Fields(string(data))
	stones := make([]int, len(stonesStr))
	for i, s := range stonesStr {
		stones[i], _ = strconv.Atoi(s)
	}
	part1 := processStonesOptimized(stones, 25)
	part2 := processStonesOptimized(stones, 75)
	fmt.Printf("Part 1: Number of stones after 25 blinks: %d\n", part1)
	fmt.Printf("Part 2: Number of stones after 75 blinks: %d\n", part2)
}

File: 2024/11/Node/main.js
const fs = require('fs');

const stones = fs.readFileSync('../input.txt', 'utf-8').trim().split(' ').map(Number);

const processStonesOptimized = (stones, blinks) => {
    let counts = new Map();

    for (const stone of stones) {
        counts.set(stone, (counts.get(stone) || 0) + 1);
    }

    for (let i = 0; i < blinks; i++) {
        let newCounts = new Map();

        for (const [stone, count] of counts.entries()) {
            if (stone === 0) {
                newCounts.set(1, (newCounts.get(1) || 0) + count);
            } else if (stone.toString().length % 2 === 0) {
                const mid = stone.toString().length / 2;
                const left = parseInt(stone.toString().slice(0, mid), 10);
                const right = parseInt(stone.toString().slice(mid), 10);
                newCounts.set(left, (newCounts.get(left) || 0) + count);
                newCounts.set(right, (newCounts.get(right) || 0) + count);
            } else {
                const newStone = stone * 2024;
                newCounts.set(newStone, (newCounts.get(newStone) || 0) + count);
            }
        }
        counts = newCounts; 
    }
    return Array.from(counts.values()).reduce((sum, count) => sum + count, 0);
};

const part1 = processStonesOptimized([...stones], 25); 
const part2 = processStonesOptimized([...stones], 75); 

console.log(`Part 1: Number of stones after 25 blinks: ${part1}`);
console.log(`Part 2: Number of stones after 75 blinks: ${part2}`);
